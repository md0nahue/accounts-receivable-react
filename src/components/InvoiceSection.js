import React from "react";

export function InvoiceSection({ items, title }) {
  if (!items?.length) {
    return <p>This section is empty!</p>;
  }

  return (
    <>
      <h5>{title}</h5>
      <table className="table">
        <thead>
          <tr>
            <th scope="col">Name</th>
            <th scope="col">Description</th>
            <th scope="col">Unit Price</th>
            <th scope="col">Quantity</th>
            <th scope="col">Discount</th>
            <th scope="col">Subtotal</th>
          </tr>
        </thead>
        <tbody>
          {items.map((item, index) => (
            <tr key={index}>
              <td>{item.name}</td>
              <td>{item.description}</td>
              <td>${(item.price_cents / 100)}</td>
              <td>{item.quantity}</td>
              <td>{item.discount_percent ? `${item.discount_percent}%` : ''}</td>
              <td>
                {item.discount_percent ? (
                  <>
                    <span className="strikethrough">
                      {`$${item.row_total_no_discount}`}
                    </span>
                    {` $${item.row_total}`}
                  </>
                ) : (
                  `$${item.row_total_no_discount}`
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  );
}
