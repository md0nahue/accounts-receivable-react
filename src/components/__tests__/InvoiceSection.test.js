import React from "react";
import { render, screen } from "@testing-library/react";
import { InvoiceSection } from "../InvoiceSection";

describe("InvoiceSection", () => {
  it("Renders a placeholder if no items prop given", () => {
    const { container } = render(<InvoiceSection />);

    expect(container.firstChild).toMatchInlineSnapshot(`
      <p>
        This section is empty!
      </p>
    `);
  });

  it("Renders a placeholder if empty items list given", () => {
    const { container } = render(<InvoiceSection items={[]} />);

    expect(screen.queryByText("This section is empty!")).not.toBeNull();
    expect(container.firstChild).toMatchInlineSnapshot(`
      <p>
        This section is empty!
      </p>
    `);
  });

  it("Does not render a placeholder if empty items list is not empty", () => {
    render(<InvoiceSection items={[1]} />);

    expect(screen.queryByText("This section is empty!")).toBeNull();
  });

  it("renders a given title as H5 heading", () => {
    render(<InvoiceSection items={[1]} title="Test Title" />);

    const heading = screen.queryByRole("heading");

    expect(heading).toBeDefined();
    expect(heading).toHaveTextContent("Test Title");
    expect(heading.tagName).toEqual("H5");
  });

  it("renders a matching snapshot", () => {
    const lineItems = [
      {
        name: "Job Many",
        description: "decribe many",
        price_cents: 10000,
        discount_percent: 10,
        quantity: 10,
      },
      {
        name: "Job Few",
        description: "describe few",
        price_cents: 10000,
        discount_percent: null,
        quantity: 1,
      },
      {
        name: "Job None",
        description: "describe none",
        price_cents: 10000,
        discount_percent: null,
        quantity: 0,
      },
    ];

    const { asFragment } = render(
      <InvoiceSection items={lineItems} title="Line Items Section" />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
