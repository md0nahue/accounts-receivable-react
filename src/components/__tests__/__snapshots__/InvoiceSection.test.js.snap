// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceSection renders a matching snapshot 1`] = `
<DocumentFragment>
  <h5>
    Line Items Section
  </h5>
  <table
    class="table"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Name
        </th>
        <th
          scope="col"
        >
          Description
        </th>
        <th
          scope="col"
        >
          Unit Price
        </th>
        <th
          scope="col"
        >
          Quantity
        </th>
        <th
          scope="col"
        >
          Discount
        </th>
        <th
          scope="col"
        >
          Subtotal
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Job Many
        </td>
        <td>
          decribe many
        </td>
        <td>
          $100
        </td>
        <td>
          10
        </td>
        <td>
          10%
        </td>
        <td>
          <span
            class="strikethrough"
          >
            $undefined
          </span>
           $undefined
        </td>
      </tr>
      <tr>
        <td>
          Job Few
        </td>
        <td>
          describe few
        </td>
        <td>
          $100
        </td>
        <td>
          1
        </td>
        <td />
        <td>
          $undefined
        </td>
      </tr>
      <tr>
        <td>
          Job None
        </td>
        <td>
          describe none
        </td>
        <td>
          $100
        </td>
        <td>
          0
        </td>
        <td />
        <td>
          $undefined
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
